Add Windows documentation

micbou opened the pull request:
This is the first draft of the official Windows documentation.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/1720)
<!-- Reviewable:end -->

Valloric commented:
Very nice! Thanks for doing this.
<!-- Reviewable comment -K09C9HAoq4I-oo0pWSr:705845434 -->

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[README.md, line 221 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09A0PODLPlyJCvXxVz)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L221)):</sup>
Try to avoid links with "here" as the anchor text. Usually the sentence can be restructured; for example:

"A copy of 64-bit Vim that's updated every week [is also available][LINK]."

or

"[A copy of 64-bit Vim that's updated every week][LINK] is also available.""

The second is better because it helps search engines provide better results.
<!-- Reviewable comment -K09A0PODLPlyJCvXxW-:627430675 -->

---

<sup>**[README.md, line 232 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09B6j_Srzs8rLm-E-n)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L232)):</sup>
Possibly add something like "If you don't know which architecture you need, you most likely need the 64 bit version."
<!-- Reviewable comment -K09B6j_Srzs8rLm-E-o:412632944 -->

---

<sup>**[README.md, line 233 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BOocjtLXFdXgMITe)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L233)):</sup>
Is this going over 80 chars?
<!-- Reviewable comment -K09BOocjtLXFdXgMITf:1375033653 -->

---

<sup>**[README.md, line 253 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BY0-DgNstsgg2TVG)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L253)):</sup>
Add "(you'll need to have `npm` installed)".
<!-- Reviewable comment -K09BY0-DgNstsgg2TVH:-1389558347 -->

---

<sup>**[README.md, line 411 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BrjtI-JnInkTB5IY)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L411)):</sup>
generator -> generators
<!-- Reviewable comment -K09BrjtI-JnInkTB5IZ:-357285731 -->

---

<sup>**[README.md, line 416 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BwCj3Nx7rFda8WuW)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L416)):</sup>
Since it probably is, maybe you should have "Win64" in the generator names and mention that they can remove it on 32bit.
<!-- Reviewable comment -K09BwCj3Nx7rFda8WuX:-1487025706 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

<sup>**[README.md, line 221 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09A0PODLPlyJCvXxVz)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L221)):</sup>
Done. I changed the weekly to frequently because it's not really updated each week.
<!-- Reviewable comment -K0CsxqOq6-XP7jdQ6EO:1498183373 -->

---

<sup>**[README.md, line 232 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09B6j_Srzs8rLm-E-n)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L232)):</sup>
Actually, the Vim obtained from the official installer is 32-bit so I am not sure it is a good idea to suggest this. Maybe I should talk about the official Vim binaries.
<!-- Reviewable comment -K0CszZrlMRzaecVt0rQ:1343264792 -->

---

<sup>**[README.md, line 233 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BOocjtLXFdXgMITe)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L233)):</sup>
Yes. I fixed this in the sections modified by the PR. There are others parts where we go over this limit but I think it should be changed in another PR.
<!-- Reviewable comment -K0CZ1Mm9Da_Y36ND2SC:-132519558 -->

---

<sup>**[README.md, line 253 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BY0-DgNstsgg2TVG)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L253)):</sup>
Done. I am also talking about `nodejs` and I updated the Mac OS X and Ubuntu installation parts. I didn't update the FreeBSD part because there are no instructions for it in the link.
<!-- Reviewable comment -K0CqRJi-UkT1yy27nSO:1407202277 -->

---

<sup>**[README.md, line 411 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BrjtI-JnInkTB5IY)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L411)):</sup>
Fixed.
<!-- Reviewable comment -K0CjMTGbRlqVST9mOkP:1758359711 -->

---

<sup>**[README.md, line 416 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BwCj3Nx7rFda8WuW)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L416)):</sup>
See my previous comment on the official Vim binaries.
<!-- Reviewable comment -K0CtnaCUu4HacDYj_TQ:792722428 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

puremourning commented:
Sorry I added my comments to the wrong file.
<!-- Reviewable comment -K0H0LT72RTibmILYH4k:98374594 -->

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[doc/youcompleteme.txt, line 379 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gyy8saAnrG5db_Jt4)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L379)):</sup>
Are we suggesting Vundle is the only supported package manager? I know that people have successfully used neobundle to have it automatically run the build for you ?
<!-- Reviewable comment -K0Gyy8taeaG_zD29-eL:-1496491450 -->

---

<sup>**[doc/youcompleteme.txt, line 386 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gz8S6PeXwEiqfxnoe)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L386)):</sup>
In this context I think "software" is plural enough

"Download and install the following software: list..."

Also, does this actually display as a list in markdown ? It looks like it is prose in raw form. I'd suggest:

```markdown
Download and install the following software:

- python2 ...
- CMake ...
- Visual Studio ...
```

Wait, I'm reviewing the wrong file. Damn this is the generated one.  In any case, it looks like `html2vimdoc` has not generated this list correctly. Maybe we can do something gin the source markdown to help it?
<!-- Reviewable comment -K0Gz8S6PeXwEiqfxnof:-1068853173 -->

---

<sup>**[doc/youcompleteme.txt, line 390 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0GzqGiBAYT4-NlzvaO)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L390)):</sup>
I know there was some discussion about this, but is 7-zip still required? I forgot the outcome of the discussion :)
<!-- Reviewable comment -K0GzqGiBAYT4-NlzvaP:2061576139 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

puremourning commented:
Also, should we add a deprecation notice to the wiki page that talked about windows installation?
<!-- Reviewable comment -K0H0_409Dd7HgHbUGeP:-524408675 -->

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[doc/youcompleteme.txt, line 379 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gyy8saAnrG5db_Jt4)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L379)):</sup>
On a side note: Vundle have changed the maintainer and is now located at https://github.com/VundleVim/Vundle.vim and IMHO we should suggest https://github.com/junegunn/vim-plug since is more maintained.
<!-- Reviewable comment -K0H1lbmi_ZFl9_IzcIZ:738026816 -->

---

<sup>**[doc/youcompleteme.txt, line 390 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0GzqGiBAYT4-NlzvaO)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L390)):</sup>
Yes, we opt for 7-zip instead of installing libclang grab what we need and uninstall it.
<!-- Reviewable comment -K0H2A95fy3dxexX5GZG:720556811 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

micbou commented:
Yes, we will do it when this PR is merged.
<!-- Reviewable comment -K0HELEgRlPn9RtOy8A_:-915974212 -->

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[doc/youcompleteme.txt, line 379 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gyy8saAnrG5db_Jt4)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L379)):</sup>
> Are we suggesting Vundle is the only supported package manager?  I know that people have successfully used neobundle to have it automatically run the build for you ?

No but I think we are recommending it because it automatically initializes and updates the submodules. NeoBundle is also able to do that. I am personnally using it. There is a section about it in the FAQ:
```vim
 NeoBundle 'Valloric/YouCompleteMe', {
     \ 'build'      : {
        \ 'mac'     : './install.py --clang-completer --system-libclang --omnisharp-completer',
        \ 'unix'    : './install.py --clang-completer --system-libclang --omnisharp-completer',
        \ 'windows' : './install.py --clang-completer --system-libclang --omnisharp-completer',
        \ 'cygwin'  : './install.py --clang-completer --system-libclang --omnisharp-completer'
        \ }
     \ }
```
but the Windows command will not work and the arguments should be removed, especially `--system-libclang` which is not recommended. I will update it.

> On a side note: Vundle have changed the maintainer and is now located at https://github.com/VundleVim/Vundle.vim and IMHO we should suggest https://github.com/junegunn/vim-plug since is more maintained.

I know there are some issues with `vim-plug` on Windows. I will try it and report here my results.
<!-- Reviewable comment -K0H1DQvoADeyBafmH1w:397512658 -->

---

<sup>**[doc/youcompleteme.txt, line 386 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gz8S6PeXwEiqfxnoe)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L386)):</sup>
I did not check the Vim help generated file. I need to find a way to fix this.
<!-- Reviewable comment -K0HFz2SRRu7mUoj_DAG:-2067997698 -->

---

<sup>**[doc/youcompleteme.txt, line 390 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0GzqGiBAYT4-NlzvaO)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L390)):</sup>
Like @vheon said, it is required for extracting the files from the Windows Clang installer without installing it.
<!-- Reviewable comment -K0H1riNqd5k2BKiYq3C:593922468 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[doc/youcompleteme.txt, line 379 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gyy8saAnrG5db_Jt4)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L379)):</sup>
I updated the `Vundle` link.
<!-- Reviewable comment -K0HMIKLcAQ6mICAx50R:-174971050 -->

---

<sup>**[doc/youcompleteme.txt, line 386 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gz8S6PeXwEiqfxnoe)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L386)):</sup>
Fixed.

> In this context I think "software" is plural enough
> "Download and install the following software: list..."

You are right about software. I didn't know it is uncountable.
<!-- Reviewable comment -K0HKzvG14YJqztUgubj:-1823381646 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[doc/youcompleteme.txt, line 379 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gyy8saAnrG5db_Jt4)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L379)):</sup>
Vundle continues to be only officially supported plugin manager for YCM. That doesn't mean we don't work with other ones, just that this one is confirmed working and we'll go _a lot_ out of our way to keep it that way.

In general, if users have several choices for something related to YCM, we officially support only one. The idea is to reduce choice paralysis and pick sound defaults.
<!-- Reviewable comment -K0I2gAL4SuUV8LaHqE0:491932975 -->

---

<sup>**[README.md, line 232 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09B6j_Srzs8rLm-E-n)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L232)):</sup>
Aren't we though suggesting the user gets a 64bit Vim a few paragraphs above? ycmd can be RAM-hungry with lots of C++ TUs and a 4GB limit with 32bit would be felt.

Maybe we should officially support only 64bit, with 32bit on a "best-effort" basis.
<!-- Reviewable comment -K0I3VQ2tkKwuDRjbkkE:877940921 -->

---

<sup>**[README.md, line 233 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09BOocjtLXFdXgMITe)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L233)):</sup>
Agreed.
<!-- Reviewable comment -K0I40O5GHHLmRedgCfV:-1530771663 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

micbou commented:
I improved the documentation by talking more about architecture.
<!-- Reviewable comment -K0M1MmEp2YzIZxIsg9b:900439715 -->

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[doc/youcompleteme.txt, line 379 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K0Gyy8saAnrG5db_Jt4)** ([raw file](https://github.com/valloric/youcompleteme/blob/bc2a3329e020d119a80ee473f6a4e27ca2d2f430/doc/youcompleteme.txt#L379)):</sup>
So I tried vim-plug on Windows and the experience is similar to Vundle. Those two plugin managers open a console window each time they execute an external command (`git clone`, `git pull`, etc.). They also freeze the Vim client during the plugins installation and update. Experience is better when using NeoBundle with [vimproc.vim](https://github.com/Shougo/vimproc.vim) and [Unite](https://github.com/Shougo/Unite.vim): no console windows and less freezing. But it is more complex to configure so we should certainly stick to Vundle.
<!-- Reviewable comment -K0I0VR-jeLQNj4AG3Nd:-1125012841 -->

---

<sup>**[README.md, line 232 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1720#-K09B6j_Srzs8rLm-E-n)** ([raw file](https://github.com/valloric/youcompleteme/blob/ae14957d0bcb91d833af65140700ddcc08ace201/README.md#L232)):</sup>
Following your comment on RAM usage, I updated the documentation by recommending the 64-bit Vim. I think we can still support the 32-bit version because we are testing it in AppVeyor.
<!-- Reviewable comment -K0M0P_7HDZ_foJlvLbc:-933995229 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

Valloric commented:
I think this is a go!

Thanks so much for working on docs; it's a thankless job that everybody finds annoying and yet it's incredibly important to users.

@homu r+
<!-- Reviewable comment -K0MkItycDIMtfWn5Qd6:-949947718 -->

---

Reviewed 2 of 3 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1720)
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 2ace4ee has been approved by `Valloric`

<!-- @homu r=Valloric 2ace4eea880cf283cc026d4f1fed07b527f15a05 -->

homu commented:
:hourglass: Testing commit 2ace4ee with merge d98298a...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/84795989)
